File: 01_Programming_Basic/Assignment/Assignment01.R
# ==========================================
# Question 1: Temperature Converter
# Goal: Practice function arguments and logic
# ==========================================

a1_q1 <- function(temperature, degree) {
  if(degree == "C"){
    answer <- (temperature - 32) * 5 / 9 # F to C
  }else if (degree == "F"){
    answer <- temperature * 9 / 5 + 32  # C to F
  }else{
    stop("Please choose 'C' or 'F' to degree. ")
  }
  return(answer)
}

# --- Test Cases ---
a1_q1(100, degree="F") # Expected: 212
a1_q1(32, degree="C")  # Expected: 0

# ==========================================
# Question 2: Vector Operations (R)
# Note: R uses 1-based indexing
# ==========================================

a1_q2 <- function(input) {
  if(length(input) != 5){
    stop("The input must include 5 elements. ")
  }else{
    if(input[5] == 0){
      stop("5th element cannot be zero. ")
    }else{
      answer <- c(input[1]+input[2], input[2]-input[3], input[3]*input[4], input[4]/input[5])
    }
  }
  return(answer)
}

# --- Test ---
input_vec <- c(10, 5, 2, 8, 4)
print(a1_q2(input_vec))
# Expected output: 15 3 16 2

# ==========================================
# Question 3: Alphabet Repeater (R)
# Note: R has built-in 'letters' vector (1-based)
# ==========================================

a1_q3 <- function(char, rep) {
  target <- letters[char]
  answer <- rep(target, times = rep) |>
    paste0(collapse = "")
  return(answer)
}

# --- Test ---
print(a1_q3(1, 5))   # [1] "aaaaa"
print(a1_q3(26, 2))  # [1] "zz"

# ==========================================
# Question 4: Max, Min, and Manual Sum (R)
# Restriction: NO sum() function allowed
# Output: A list with names: max, min, sum
# ==========================================

a1_q4 <- function(numbers) {
  max_value <- max(numbers)
  min_value <- min(numbers)
  sum_value <- 0
  for(i in seq_len(length(numbers))){
    sum_value <- sum_value + numbers[i]
  }
  answer <- list(max = max_value, min = min_value, sum = sum_value)
  return(answer)
}

# --- Test ---
my_vec <- c(1, 2, 3, 4, 5)
print(a1_q4(my_vec))
# Expected output: $max [1] 5  $min [1] 1  $sum [1] 15

# ==========================================
# Question 5: Even/Odd Filter (R)
# Logic: Vectorized filtering is more R-like
# ==========================================

a1_q5 <- function(lower, upper, is_even) {
  answer <- c()
  numbers <- ceiling(lower):floor(upper)
  for(i in seq_along(numbers)){
    if(is_even){
      if(numbers[i] %% 2 == 0){
        answer <- c(answer, numbers[i])
      }
    }else{
      if(numbers[i] %% 2 != 0){
        answer <- c(answer, numbers[i])
      }
    }
  }
  return(answer)
}

# --- Test ---
print(a1_q5(lower=11, upper=22, is_even=TRUE))
# Expected output: 12 14 16 18 20 22

File: 01_Programming_Basic/Assignment/Assignment01.py
# =====================================================
# Question 1: Temperature Converter (Python version)
# Function: a1_q1(value, degree)
# Description: Convert between Celsius and Fahrenheit
# =====================================================

import math

def a1_q1(temperature, degree):
    if degree == "C":
        answer = (temperature - 32) * 5 / 9  # F to C
    elif degree == "F":
        answer =  temperature * 9 / 5 + 32  # C to F
    else:
        print("Please choose 'C' or 'F' for degree. ")
    return int(answer)

# --- Test Examples ---
# Example 1: 100°C -> 212°F
print(a1_q1(100, degree="F")) 

# Example 2: 32°F -> 0°C
print(a1_q1(32, degree="C"))

# =====================================================
# Question 2: List Operations (Python version)
# Input: A list with 5 numbers
# Output: A list with 4 calculated terms
# =====================================================

def a1_q2(input):
    if len(input) != 5:
        print("Error: input list must have exactly 5 elements. ")
    else:
        if input[4] == 0:
            print("Error: 5th element is zero. ")
        else:
            answer = [round(input[0]+input[1], 10), round(input[1]-input[2], 10), round(input[2]*input[3], 10), round(input[3]/input[4], 10)]
    return answer

#--- Test Examples ---
input_list = [10, 5, 2, 8, 4]
print(a1_q2(input_list))
# expected output: [15, 3, 16, 2.0]

# =====================================================
# Question 3: Alphabet Repeater (Python version)
# =====================================================

def a1_q3(char, rep):
    import string as st
    target = st.ascii_lowercase[char - 1]
    answer = target * rep
    return answer

# --- Test ---
print(a1_q3(1, 5))   # expected output: 'aaaaa'
print(a1_q3(3, 3))   # expected output: 'ccc'

# =======================================================
# Question 4: Max, Min, and Manual Sum (Python version)
# =======================================================

def a1_q4(numbers):
    max_num = max(numbers)
    min_num = min(numbers)
    sum_number = 0
    for i in range(len(numbers)):
        sum_number += numbers[i]
    answer = {"max": max_num, "min": min_num, "sum": sum_number}
    return answer

# --- Test ---
my_list = [1, 2, 3, 4, 5]
print(a1_q4(my_list))
# expected output: {'max': 5, 'min': 1, 'sum': 15}

# ================================================
# Question 5: Even/Odd Flitter (Python version)
# ================================================

def a1_q5(lower, upper, is_even):
    answer = []
    numbers = list(range(int(lower), int(upper)+1))
    for i in numbers:
        if is_even:
            if i % 2 == 0:
                answer.append(i)
        else:
            if i % 2 != 0:
                answer.append(i)
    return answer

# --- Test ---
print(a1_q5(lower=11, upper=22, is_even=True))
# expected output: [12, 14, 16, 18, 20, 22]
